packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -646,22 +646,43 @@ export default createRule<Options, MessageId>({
           .getCallSignaturesOfType(
             getConstrainedTypeAtLocation(services, callback),
           )
-          .map(sig => sig.getReturnType());
-        /* istanbul ignore if */ if (returnTypes.length === 0) {
-          // Not a callable function
-          return;
-        }
-        // Predicate is always necessary if it involves `any` or `unknown`
-        if (returnTypes.some(t => isTypeAnyType(t) || isTypeUnknownType(t))) {
-          return;
+          .map(sig => sig.getReturnType())
+          .map(t => {
+            // TODO: use `getConstraintTypeInfoAtLocation` once it's merged
+            // https://github.com/typescript-eslint/typescript-eslint/pull/10496
+            if (tsutils.isTypeParameter(t)) {
+              return checker.getBaseConstraintOfType(t);
+            }
+
+            return t;
+          });
+
+        let hasFalsyReturnTypes = false;
+        let hasTruthyReturnTypes = false;
```

Comment:
One more optimization: we know that we won't care about possibly truthy types if there was a possibly falsy type. So this can become:

```suggestion
        let knownReturnType: 'falsy' | 'truthy' | undefined;
```

...and the truthiness check below could be:

```ts
} else if (knownReturnType !== 'falsy' && isPossiblyTruthy(type)) {
  knownReturnType = 'truthy';
}
```

Reply:
I like the suggested optimization, but I don't think the proposed change matches the current implementation (some tests fail unless I've missed something).

To my understanding, the current code can be changed to the following so that as soon as both a truthy and a falsy values exist, the function bails:

```diff
for (const type of returnTypes) {
  // Predicate is always necessary if it involves `any` or `unknown`
  if (!type || isTypeAnyType(type) || isTypeUnknownType(type)) {
    return;
  }

  if (isPossiblyFalsy(type)) {
    hasFalsyReturnTypes = true;
  }

  if (isPossiblyTruthy(type)) {
    hasTruthyReturnTypes = true;
  }

+  if (hasFalsyReturnTypes && hasTruthyReturnTypes) {
+    return;
+  }
}
```

I think the same value can be possibly-truthy and possibly-falsy at the same time (e.g. `boolean`), so both checks have to run for every return type.

I think matching the current implementation with a single value state (`let knownReturnType: 'falsy' | 'truthy' | undefined`) will require potentially nested if statements, as even something like the following fails to catch some cases (like the returned `boolean` type):

```ts
if (isPossiblyFalsy(type) && knownReturnType !== 'truthy') {
  knownReturnType = 'falsy';
}

if (isPossiblyTruthy(type) && knownReturnType !== 'falsy') {
  knownReturnType = 'truthy';
}
```

I'll be happy to update with the above diff if it makes sense or something different if I'm missing something.

Reply:
Gotcha - that works!